lib/typing_env.ml

```diff
@@ -1,50 +1,121 @@
 open Types
 
-let extract_non_alias_types signature =
+type t = { env : Env.t; subst : Subst.t }
+type subst_kind = Type | Module | Modtype [@@deriving ord]
+
+module Subst_item_map = Map.Make (struct
+  type t = subst_kind * string [@@deriving ord]
+end)
+
+let uniq_signature_ids ~reference ~current =
```

Comment:
There a quite a lot of substitution going on so I think providing a bit of context for this function would help.

I am correct that this will traverse the whole `current` sig, registering a subst for any ID that already exist in the `reference` signature to a fresh ID and then running the substitution to return a new signature with IDs updated?

I think we could both have a better name for this function and document it to describe this behaviour in a few lines so we don't have to read the code to understand what it does.

How about `replace_matching_ids ~reference ~current` or something along those lines?

Reply:
> I am correct that this will traverse the whole current sig, registering a subst for any ID that already exist in the reference signature to a fresh ID and then running the substitution to return a new signature with IDs updated?

Exactly!